### PR TITLE
File context menu improvements

### LINKED ISSUE #2950: Add the ability to "copy path to file" and "open in VSCode" on file right click

### PR DESCRIPTION
**Pull request by Pranav2612000**
Fixes #2950

### CODE REVIEW

**Comment by Pranav2612000** [comment]
For `Open in VSCode` & `Copy Absolute Path` we'll need the base project directory location. I couldn't find a way to get this directly.

One solution is to pass the project ( or `projectService` ) as props to the `FileContextMenu` component, but that would involve a fair bit of prop-drilling.
The other option is to add an `absolutePath` property to the `LocalFile` type we have.

@krlvi What do you think should we do? Or is there a different better approach?

**Comment by mtsgrd** [comment]
You're right about prop drilling, but that is how we have been doing things to date. What I'd prefer is some form of automated dependency injection the way Angular does it, but Svelte's `setContext` and `getContext` methods provide no lifecycle management. Currently what this means is when you e.g. switch between projects components that are recycled rather than destroyed and recreated would not be given the latest dependencies created in the `layout.ts` files.

So we do need a solution here, but until we have a plan it feels prop drilling is better than e.g. going with global singletons. What do you think?

**Comment by Pranav2612000** [comment]
Makes sense @mtsgrd
I was also thinking we need a global state management tool like React Context.
I'm not very well versed but with Svelte, but I get the general idea of what you are trying to say. 
I think the stale context would be a common issue. Do you have an idea of how others do it? I'll be happy to spend more time and see if there are any workarounds we can come up with.

In the meantime, I'll update this PR to pass the props to the `FileContextMenu` component.
Thank you so much for your help.

**Comment by Pranav2612000** [comment]
Updated the PR to use prop drilling. Feel free to review

File: gitbutler-ui/src/lib/components/FileContextMenu.svelte [hunk 1]
```diff
@@ -29,13 +33,57 @@
...
+							}
+						}}
+					/>
+					<ContextMenuItem
+						label="Copy Relative Path"
+						on:click={() => {
+							try {
+								if (!project) {
+									return;
+								}
```
Comment (mtsgrd) [inline gitbutler-ui/src/lib/components/FileContextMenu.svelte]: For consistency with the code base I'd recommend expressing this as: `if(!project) return;`

File: gitbutler-ui/src/lib/components/FileContextMenu.svelte [hunk 2]
```diff
@@ -29,13 +33,57 @@
...
+								navigator.clipboard.writeText(item.files[0].path);
+								dismiss();
+							} catch (err) {
+								console.log('Failed to copy relative path', err);
```
Comment (mtsgrd) [inline gitbutler-ui/src/lib/components/FileContextMenu.svelte:67]: I think we want to show a toast here, could you add something like this?

```
import * as toasts from '$lib/utils/toasts';
toasts.error('Failed to copy relative path', err);
```

**Comment by Pranav2612000** [comment]
@mtsgrd I've updated the code to add error toasts. Can you please take another look

File: gitbutler-ui/src/lib/components/FileContextMenu.svelte [hunk 3]
```diff
@@ -29,13 +34,58 @@
...
+					on:click={async () => {
+						try {
+							if (!project) return;
+							for (let file of item.files) {
+								const absPath = await join(project.path, file.path);
+								open(`vscode://file${absPath}`);
+							}
+							dismiss();
+						} catch {
+							console.log('Failed to open in VSCode');
```
Comment (mtsgrd) [inline gitbutler-ui/src/lib/components/FileContextMenu.svelte:84]: Can we change this to  catch the error, switch log to `console.error()` and pass the err as a second argument?

File: gitbutler-ui/src/lib/components/FileContextMenu.svelte [hunk 4]
```diff
@@ -29,13 +34,58 @@
...
+					/>
+					<ContextMenuItem
+						label="Copy Relative Path"
+						on:click={() => {
+							try {
+								if (!project) return;
+								navigator.clipboard.writeText(item.files[0].path);
+								dismiss();
+							} catch (err) {
+								console.log('Failed to copy relative path', err);
```
Comment (mtsgrd) [inline gitbutler-ui/src/lib/components/FileContextMenu.svelte:67]: Same here, let's use `console.error()`?

**Review by mtsgrd — APPROVED** [review]

**Comment by mtsgrd** [comment]
Added a couple of comments, but have approved so feel free to merge when ready!

**Comment by krlvi** [comment]
Thanks for the contribution! I am gonna merge this and change the logs to `console.error()` in a separate PR

**Pull request merged**